README.md

```diff
@@ -29,95 +33,319 @@ are well established for RSE-ops, and which require further development. We can
+extreme parallelism [^8], it's relatively trivial to scale a task on an HPC cluster. However, the user has to know what they are doing.
+
+> Don't these two sentences contradict each other?
+
+A related issue is testing at scale. As tests need to be run frequently and across many different architectures, the incentive structure is arguably not there to maintain such a testing framework.
 
+> we really need an HPC test harness. There are attempts at this for app testing (3 versions of a
+> tool called ATS internally) and there ar things like Pavilion2
+> (https://github.com/hpc/pavilion2), ReFrame (https://github.com/eth-cscs/reframe) and BuildTest
+> (1 developer/user who you know) for facility testing but there isn't anything I can see here that
+> is converged or standard. Someone needs to build a test tool for not just their own project.
+
+This is in contrast to some Cloud Native tools that handle scaling or scaled testing for the user. This suggests that the greatest challenges are not related to the technologies, but educating the user base and RSEs how to properly use them. There is a huge opportunity for development of RSE-ops tools that can better automate this scaling.
+
+> yes
 
 ### What about software distribution?
 
-The end-goal of an HPC environment is to make software available to researchers. This is abstractly similar to the "end goal" of a cloud native environment, which is to make software available to users via services. For cloud native, I suspect this typically means interactive web interfaces or services that are accessed from other web or terminal clients via application programming interfaces (APIs). Since HPC environments are typically running jobs that rely on system software and some number of known aliases, the bulk majority of "software distribution" for HPC is probably command line interfaces. In this framework, the administrators must provide an easy way to manage a collection of software, and make it available to users. 
+The end-goal of an HPC environment is to make software available to researchers. This is abstractly similar to the "end goal" of a cloud native environment, which is to make software available to users via services. For cloud native, I suspect this typically means interactive web interfaces or services that are accessed from other web or terminal clients via application programming interfaces (APIs). Since HPC environments are typically running jobs that rely on system software and some number of known aliases, the bulk majority of "software distribution" for HPC is probably command line interfaces. In this framework, the administrators must provide an easy way to manage a collection of software, and make it available to users.
+
+> Again here it's not just admins. App teams maintain their own app deployments. The end result of
```

Comment:
I'm not actually familiar even with what an "app" is on an HPC cluster... I need to engage with the code teams to see!

Reply:
Simple example with a bit of info on the code it came from: https://asc.llnl.gov/codes/proxy-apps/lulesh

The app is an executable, typically takes some type of input file, usually in a (sometimes custom) scripting language.  The script drives some time-stepped physics simulation.  Code runs, outputs data about timesteps / progress of the simulation, periodically dumps files for:

- restart (in case it crashes -- these are called "checkpoints")
- visualization (to look at what's going on)
- various interesting summary stats/results -- maybe a mesh transformed by whatever happened.

You generally `srun` or `mpirun` the code, and you may need to configure things like where do these dumps go (as they can be quite large) -- i.e. you want to put some of them in the parallel FS.

Executable might link to a lot of libraries -- math libs, other code components, etc.